### PR TITLE
Add coarse module profiler instrumentation

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #3974

An example with this 

image

You can see "blockblock::layers.23", "attention::layers.23._checkpoint_wrapped_module.attention, "moe::layers.23._checkpoint_wrapped_module.moe"

Summary:
- Add module-level profiler instrumentation for coarse pretraining regions such as transformer blocks, attention, FFN, and MoE modules.
- Emit `record_function` contexts and best-effort `mark_kernels` metadata from the same wrapper under `--profiler.enable_module_profiler`.
- Wire the feature through `Profiler.Config` and keep it disabled by default.
- Document how to enable CPU/GPU profiling and module profiler annotations.

### CODE REVIEW

**Review by fegin — COMMENTED** [review]
What's the issue if we always apply the record_function? Is CPU overhead the main concern?

**Comment by anijain2305** [comment]
@fegin it is now enabled, and the PR is ready for review.

File: torchtitan/tools/module_profiler.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    if len(parts) >= 2 and parts[-2] == "layers" and name.isdigit():
+        return "block"
+    if name in {"attention", "attn"} or name.endswith("_attn"):
+        return "attention"
+    if name == "feed_forward":
+        return "ffn"
+    if name == "moe":
+        return "moe"
+    return None
```
Comment (tianyu-l) [inline torchtitan/tools/module_profiler.py]: things like this is creating maintaining difficulty

File: torchtitan/tools/module_profiler.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        _args: tuple[Any, ...],
+        _kwargs: dict[str, Any],
+        _output: Any,
+    ) -> None:
+        if stacks:
+            stacks.pop().close()
+
+    module.register_forward_pre_hook(pre_hook, prepend=True, with_kwargs=True)
+    module.register_forward_hook(post_hook, with_kwargs=True, always_call=True)
+    setattr(module, _WRAPPED_ATTR, True)
```
Comment (tianyu-l) [inline torchtitan/tools/module_profiler.py]: how does this compare with native solution in https://docs.pytorch.org/docs/2.13/profiler.html

> with_modules ([bool](https://docs.python.org/3/library/functions.html#bool)) – record module hierarchy (including function names) corresponding to the callstack of the op. e.g. If module A’s forward call’s module B’s forward which contains an aten::add op, then aten::add’s module hierarchy is A.B Note that this support exist, at the moment, only for TorchScript models and not eager mode models.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/tools/module_profiler.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                module_kind=kind,
+                model_part_idx=model_part_idx,
+            ):
+                num_wrapped += 1
+
+    logger.info(f"Applied module profiler instrumentation to {num_wrapped} module(s)")
+    return num_wrapped
+
+
+def coarse_module_kind(module_fqn: str) -> str | None:
```
Comment (sanketpurandare) [inline torchtitan/tools/module_profiler.py]: How bad is it to use module_fqn directly?

File: torchtitan/tools/module_profiler.py (hunk 1 above)
Comment (anijain2305) [reply torchtitan/tools/module_profiler.py]: changed this to just use module_fqns. Also restricting the profile to only a select few modules for now

```
isinstance(module, (TransformerBlock, BaseAttention, FeedForward, MoE))
```

to prevent large noise.

File: torchtitan/tools/module_profiler.py (hunk 3 above)
Comment (anijain2305) [reply torchtitan/tools/module_profiler.py]: Changed to module_fqn

File: torchtitan/tools/module_profiler.py (hunk 2 above)
Comment (anijain2305) [reply torchtitan/tools/module_profiler.py]: `with_modules` is only for Torchscript, not for eager.

File: tests/unit_tests/test_module_profiler.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+from torchtitan.models.common.attention import BaseAttention
+from torchtitan.models.common.decoder import TransformerBlock
+from torchtitan.models.common.feed_forward import FeedForward
+from torchtitan.models.common.moe import MoE
+
+from torchtitan.tools import module_profiler
+from torchtitan.tools.module_profiler import apply_module_profiler
+
+
+class _Attention(BaseAttention):
```
Comment (sanketpurandare) [inline tests/unit_tests/test_module_profiler.py]: Why can't we directly use `TransformerBlock` with some debug config?

File: torchtitan/tools/module_profiler.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
```
Comment (sanketpurandare) [inline torchtitan/tools/module_profiler.py]: Can the module_profiler be consolidated as a part of torchtitan.tools.profiler.Profiler?

**Comment by anijain2305** [comment]
@sanketpurandare addressed the feedback

**Comment by anijain2305** [comment]
@sanketpurandare @tianyu-l ping

File: torchtitan/tools/profiler.py [hunk 6]
```diff
@@ -256,6 +266,151 @@ def step(self) -> None:
             with sl.log_trace_span("memory_profiler_step_call"):
                 self.memory_profiler.step()
 
+    @classmethod
+    def apply_module_profiler(
```
Comment (tianyu-l) [inline torchtitan/tools/profiler.py:270]: This reminds me of https://github.com/pytorch/torchtitan/blob/main/agent_tooling/numerics_debugging/activation_tracer.py#L502

I feel this either belongs to PyTorch core profiler, or this `agent_tooling` folder (using skills and agents to obtain something useful, but code doesn't need to be maintained in core).

Comment (anijain2305) [reply torchtitan/tools/profiler.py:270]: Hmm, I think this belonging to core sounds questionable to me.

This code is very much specialized to how TorchTitan models are written, and to overcome some of the hiccups that I saw while profiling titan models. For example, I had to install this profiler as a pre and post-install hook (instead of a context manager as advertised in profiler docs) otherwise the profiler boxes were not present in a nice parent-child relationship in the perfetto trace.

There was some navigation around the SAC calling TorchDispatchMode because of which `record_function` was not working out perfectly. Relevant comments are


        # The public torch.profiler.record_function enters and exits through
        # dispatcher ops. Selective activation checkpointing installs a
        # TorchDispatchMode, so those profiler ops also show up as
        # PythonDispatchMode events in raw traces. Those PythonDispatchMode
        # events can start before a module range exits and finish after it,
        # which makes Perfetto display the module ranges as crossing siblings
        # instead of cleanly nested scopes.
        #
        # _RecordFunctionFast opens the same user annotation directly in the
        # C++ profiler path without dispatching an op, so the raw trace keeps
        # the module ranges properly nested under TorchDispatchMode.


Similarly I think agent_tooling sounds like its only for agents. But atleast as of now, there are still performance engineers who like to look at traces.

**Comment by felipemello1** [comment]
hey @anijain2305, 

1. do you have an image of how it looks **before** your PR?
2. is this `with_stacks=True`?
3. Have you had a chance to see if megatron/llama4x has similar problem/solution?
4. And just for my own understanding: You have to use hooks because, if you use context manager `record_function`, theres a race condition and the trace is not computed correctly. Is that so? It sounds to me like a profiler issue. Do you agree, or do you think thats because of how TorchTitan does things? I am saying this based on your comment:

 "There was some navigation around the SAC calling TorchDispatchMode because of which record_function was not working out perfectly."
 
 If its a profiler issue, i would say we should file a bug.

cc: @scotts

**Review by sanketpurandare — APPROVED** [review]
LGTM!

**Comment by sanketpurandare** [comment]
> 2. is this `with_stacks=True`?

with_stacks=True adds a lot of profiler overhead.

**Comment by scotts** [comment]
From the perspective of working on the profiler, I suspect this is not something that Titan will want to maintain. If the profiler is not behaving in an expected way, let's drill down into that issue and try to fix that.

**Comment by anijain2305** [comment]
@felipemello1 Thanks for the review



> 1. do you have an image of how it looks **before** your PR?


Yes, this is what it looks like. 

**Baseline (no instrumentation)**

This is the [trace](https://www.internalfb.com/intern/perfetto/open_trace/?manifold_path=perfetto_internal_traces%2Ftree%2Fshared_trace%2Fanijain_d8632fb9-1376-41c5-b387-dde220c018a5_rank0_trace.json.gz).
 
There is no way to easily tell that MoE is taking 80% of the transformer time. There is no easy way for me to tell where is the all_gather of async-TP.

image


**Attempt 1 - Lets just add `with record_function()`** 

This is the [trace](https://www.internalfb.com/intern/perfetto/open_trace/?manifold_path=perfetto_internal_traces%2Ftree%2Fshared_trace%2Fanijain_3c723bcd-20d5-42b6-8ef5-fede94f1a0c7_rank0_trace.json.gz).

As you can see, there are boxes but they are not lined up well. The record function event has to registered before the first kernel call, otherwise the parent-child relationship is not nicely represented in the perfetto traces.

image



**Attempt 2 - RecordFunctionFast**

This is the [trace](https://www.internalfb.com/intern/perfetto/open_trace/?manifold_path=perfetto_internal_traces%2Ftree%2Fshared_trace%2Fanijain_7571b67d-11e8-40dc-89bb-29b87a0dabb7_rank0_trace.json.gz).

image

Look at the red rectangle, it shows the `attention` and `feed_forward` block. At this point, I am happy because I can see main transformer components line up nicely.



> 2. is this `with_stacks=True`?

No, this is not using `with_stacks=True`. This is using explicit module ranges via record_function. `with_stacks` is orthogonal and would add stack collection overhead.

> 3. Have you had a chance to see if megatron/llama4x has similar problem/solution?

No, I have not looked into megatron/llama4x. But from first principles, it is definitely desirable to have boxes. This is very common in Ads models. You should be able to look at trace and say - "Hey, MoE is taking 90% of the time, I did not expect that!".


> 4. And just for my own understanding: You have to use hooks because, if you use context manager `record_function`, theres a race condition and the trace is not computed correctly. Is that so? It sounds to me like a profiler issue. Do you agree, or do you think thats because of how TorchTitan does things? I am saying this based on your comment:
> 
> "There was some navigation around the SAC calling TorchDispatchMode because of which record_function was not working out perfectly."
> 
> If its a profiler issue, i would say we should file a bug.
> 

It is not really a race condition. There are two separate issues:

* We use hooks to install ranges automatically around TorchTitan module boundaries without changing each model’s forward code. We could have done monkeypatching like

```
orig_call = module.__call__

def new_call():
    with record_function():
           orig_call()

module.__call__ = new_call
```

But this is not pythonic. Pre-/post-hooks are pythonic. 


* Secondly, now one could say, why not use `record_function.__enter__` and `record_function.__exit__` in these pre-/post-hooks. This is where the dispatcher problem comes in. The public `torch.profiler.record_function` enters/exits through dispatcher ops. Under selective activation checkpointing, TorchDispatchMode sees those profiler ops, which creates `PythonDispatchMode` slices that cross the intended module range boundaries in raw traces (this is the attempt 1 snapshot earlier).`_RecordFunctionFast` avoids the dispatcher path, so the raw trace nesting stays clean. This is the zoomed-in snapshot where the `PythonDispatchMode` slice starts earlier, preventing the nesting

image

I agree that the profiler team might improve the `PythonDispatchMode` slice interaction but this is an interaction that happens with SAC, record_function and how titan models use them. But this is not a profiler bug.


cc @scotts for comments.

**Comment by anijain2305** [comment]
@scotts Can you comment more on the maintenance burden side? At a high-level, this PR is just adding record function labels for the common Transformer model subcomponents. We dont want to annotate everything to prevent  visual noise in the perfetto traces (like we just want to see attention, MoE etc, but not every linear call with a box).

Imo, the only thing that could be questionable is the use of `RecordFunctionFast` and for that I describe the issue more in https://github.com/pytorch/torchtitan/pull/3974#issuecomment-5146912408. If this can be fixed, I am happy to go back to `record_function.__enter__` and `record_function.__exit__`. But I doubt if thats what you meant by maintenance burden.

Edit - My agent was able to repro the SAC + record_function issue here (for which I used RecordFunctionFast as the workaround in this PR) - https://github.com/pytorch/pytorch/issues/191797

File: torchtitan/tools/profiler.py [hunk 7]
```diff
@@ -256,6 +266,151 @@ def step(self) -> None:
...
+        module_fqn: str,
+        model_part_idx: int,
+    ) -> bool:
+        if getattr(module, cls.MODULE_PROFILER_WRAPPED_ATTR, False):
+            return False
+
+        label = module_fqn
+        annotation: dict[str, Any] = {
+            "name": label,
+            "module_fqn": module_fqn,
```
Comment (felipemello1) [inline torchtitan/tools/profiler.py:332]: Just double checking, is repeated name module_fqn intended?

**Review by felipemello1 — APPROVED** [review]
Thanks for the explanation! LGTM! My 2c: I would wait for @tianyu-l and @scotts comments before merging

File: torchtitan/tools/profiler.py [hunk 8]
```diff
@@ -256,6 +266,151 @@ def step(self) -> None:
...
+        return num_wrapped
+
+    @staticmethod
+    def should_profile_module(module: nn.Module) -> bool:
+        from torchtitan.models.common.attention import BaseAttention
+        from torchtitan.models.common.decoder import TransformerBlock
+        from torchtitan.models.common.feed_forward import FeedForward
+        from torchtitan.models.common.moe import MoE
+
+        return isinstance(module, (TransformerBlock, BaseAttention, FeedForward, MoE))
```
Comment (tianyu-l) [inline torchtitan/tools/profiler.py:316]: what's the criteria of having these four modules, not others?

this creates maintenance burden -- one could easily modify / add modules (e.g. linear attention) and forget to edit this list

File: torchtitan/trainer.py [hunk 9]
```diff
@@ -437,6 +437,9 @@ def __init__(self, config: Config):
 
                 self.model_parts = [model]
 
+        if config.profiler.enable_profiling:
+            Profiler.apply_module_profiler(self.model_parts)
```
Comment (tianyu-l) [inline torchtitan/trainer.py:441]: this is hacky, alternative way is to send `self.model_parts` into the constructor of profiler
https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L952

File: torchtitan/tools/profiler.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/tools/profiler.py:270]: > This code is very much specialized to how TorchTitan models are written, and to overcome some of the hiccups that I saw while profiling titan models.

I'm not arguing that the current code is not titan-specific. I'm wondering if PyTorch profiler should provide better tooling / UX for any users so that they don't need to do model-specific / framework-specifc treatment.

File: torchtitan/tools/profiler.py (hunk 8 above)
Comment (anijain2305) [reply torchtitan/tools/profiler.py:316]: I chose these because these are kind of key components in transformer models. The usual practice while profiling is to add `record_function` in the model code around the interesting parts that you want to see in the profiler traces.

If there are some base classes that we are likely to always derive from, that can be a good starting point. 

I can also refactor this to be programmable so that users can provide their own list of modules. But we **do need** some defaults, and therefore some maintenance burden.

Stepping back, I think we should establish if there is value in this PR. The main difference looks like this

Today
image

With coarse grained labelling 
image

If there is value, then the next thing to answer is - how much maintenance in titan vs core? I don't think Pytorch profiler provides a way (and I don't think it should because it should not take an `nn.module` as an input) to label all or even select set of modules. 

However if the maintenance outweighs the value of this PR, I am ok rewriting this as instruction for agents (as you suggested earlier).

File: torchtitan/trainer.py (hunk 9 above)
Comment (anijain2305) [reply torchtitan/trainer.py:441]: This is definitely better. I can try after this comment is discussed - https://github.com/pytorch/torchtitan/pull/3974#discussion_r3694734206

File: torchtitan/tools/profiler.py (hunk 8 above)
Comment (scotts) [reply torchtitan/tools/profiler.py:316]: @anijain2305, echoing @tianyu-l , I also see this as a maintenance burden. This is also the core mechanism of the PR: you want these modules to be treated differently than all of the rest, because they tend to have special meaning to your users. My concern is both that this list is going to grow, and that it's going to start getting special logic that users will need options to control ("Don't include MoE if X"). Over time, that's tends to get ugly.

Comment (tianyu-l) [reply torchtitan/tools/profiler.py:316]: @anijain2305 
> The usual practice while profiling is to add `record_function` in the model code around the interesting parts that you want to see in the profiler traces.

why don't we go with the usual practice?

Comment (fegin) [reply torchtitan/tools/profiler.py:316]: @anijain2305 Here is my proposal:

1. Add one field, `fqn`, to `Module`: I think this is going to be used a lot. For example @felipemello1's logging may need FQN as well. It is easy to populate fqn for each Module. We can do that in `parallelize()` since we do recursively traversal anyway. Be careful to skip checkpoint wrapper.
2. Add one field, `should_record`, to `Module.Config`: the users can decide what to turn on by tuning the configuration.
